orttraining/orttraining/training_ops/cuda/torch/torch_custom_function_kernel.cc

```diff
@@ -0,0 +1,307 @@
+  AddFloatTupleArgs();
+  AddPointerScalarArgs();
+}
+
+void PythonOp::CreateArgPositions() {
+  ORT_ENFORCE(arg_positions_.size() == 0);
+
+  // occupied[i] being true means the i-th input argument
+  // to Python function has been set.
+  std::vector<bool> occupied(input_tensor_types_.size() + const_args_.size(), false);
+
+  // We know all non-tensors were set above, so let's catch up.
+  for (const auto pos : const_arg_positions_) {
+    occupied.at(pos) = true;
+  }
+
+  // Search for empty slots for tensors.
+  // The i-th empty slot is assigned the i-th input tensor.
+  for (size_t i = 0; i < occupied.size(); ++i) {
+    if (occupied.at(i)) {
+      continue;
+    }
+    // arg_positions[i] is the position index for the i-th tensor input.
+    // Find an empty slot whose position index is i.
+    arg_positions_.push_back(i);
+  }
+}
+
+std::vector<OrtValue*> CreateArgs(OpKernelContext* context, const size_t begin_index) {
+  auto* ctx_internal = reinterpret_cast<onnxruntime::OpKernelContextInternal*>(context);
+  std::vector<OrtValue*> args;
+  for (size_t i = begin_index; i < static_cast<size_t>(ctx_internal->InputCount()); ++i) {
+    args.push_back(const_cast<OrtValue*>(ctx_internal->GetInputMLValue(i)));
+  }
+  return args;
+}
+
+void SetContextOutput(OpKernelContext* context, std::vector<void*>& returned_args) {
+  // Handle the outputs;
+  // The 1st output is context index of auto grad function.
+  // Other outputs are address of OrtValue we got from python script run.
+  PyObject* ctx_addr = reinterpret_cast<PyObject*>(returned_args[0]);
+  ORT_ENFORCE(ctx_addr, "Context object pointer should not be null");
+
+  // todo(pengwa): optional? re-visit this once we have better understanding on how PyTorch handle the ref cnt.
+  Py_INCREF(ctx_addr);
+
+  Tensor* ctx_id_tensor = context->Output(0, {1});
+  ORT_ENFORCE(ctx_id_tensor != nullptr, "Context tensor should not be null.");
+  int64_t* ctx_id_data = ctx_id_tensor->template MutableData<int64_t>();
+  *ctx_id_data = onnxruntime::language_interop_ops::torch::OrtTorchFunctionPool::GetInstance().RegisterContext(ctx_addr);
+}
+
+void SetOtherOutputs(OpKernelContext* context, std::vector<void*>& returned_args) {
+  auto* ctx_internal = reinterpret_cast<onnxruntime::OpKernelContextInternal*>(context);
+  for (size_t i = 1; i < static_cast<size_t>(ctx_internal->OutputCount()); ++i) {
+    // The returned pointer points to a OrtValue created on Python side.
+    // Here we just cast it to the right type.
+    OrtValue* ptr = reinterpret_cast<OrtValue*>(returned_args.at(i));
+    ORT_ENFORCE(ptr, "Returned argument from Python should not be null.");
+    // Set the output directly to Python-generated OrtValue value.
+    ORT_THROW_IF_ERROR(ctx_internal->SetOutputMLValue(i, *ptr));
+  }
+}
+
+Status PythonOp::ComputeInternal(OpKernelContext* context) const {
+  // Todo(pengwa): perf impact and how much, leave it now to guarantee correctness.
+  CUDA_RETURN_IF_ERROR(cudaDeviceSynchronize());
```

Comment:
Would be nice to identify if we need this for correctness, or if we can rely on ORTModule launching kernels on the same stream that PyTorch will use.  I think these synchronization steps are the only CUDA-specific pieces of work here.

Reply:
User can launch functions on other GPUs in a custom functions. I think for computation, it's 99.999% safe to remove it. But I am not sure what kind of communication user would do (e.g., a sub-group reduction for Megatron), so I perfer to keep it.